packages/stark-ui/src/modules/date-picker/components/date-picker.component.ts

```diff
@@ -229,40 +247,52 @@ export class StarkDatePickerComponent extends AbstractStarkUiComponent
 	public dateMaskConfig?: StarkTimestampMaskConfig = undefined;
 
 	/**
+	 * Part of {@link MatFormFieldControl} API
 	 * @ignore
 	 * @internal
 	 */
 	// tslint:disable-next-line:no-null-keyword
 	public ngControl: NgControl | null = null;
 
 	/**
-	 * Variable to define to use MatFormFieldControl.
-	 * Stream that emits whenever the state of the control changes such that the parent `MatFormField`
-	 * needs to run change detection.
+	 * Part of {@link MatFormFieldControl} API
+	 * @ignore
+	 * @internal
 	 */
 	public stateChanges: Subject<void> = new Subject<void>();
 
 	/**
+	 * Part of {@link MatFormFieldControl} API
 	 * @ignore
 	 * @internal
 	 */
 	public focused = false;
 
 	/**
-	 * Variable to define to use MatFormFieldControl
-	 * Whether the control is in an error state.
+	 * @ignore
+	 * @internal
+	 */
+	public pickerInputTouched = false;
+
+	/**
+	 * Part of {@link MatFormFieldControl} API
+	 * @ignore
+	 * @internal
 	 */
 	public get errorState(): boolean {
+		// the control can be in an error state as long as the user has interacted with it
 		return (
 			this.ngControl !== null &&
 			this.ngControl.control !== null &&
+			this.pickerInputTouched &&
```

Comment:
Now, we are checking if it is **touched** which is better than before.
I think we should also check if it is **dirty** or not. Adding this condition would be probably better:
`(this.pickerInputTouched || this.pickerInputDirty) &&`

Othewise it won't be marked as error when the user submit the form...

Reply:
I agree... but there is a problem though, submitting the form does not mark the control as `dirty`.

So in this case, it is the developer that should mark the control as `dirty` or `touched`. Moreover there is no way (in the Angular API nor Material API) to know when then control has changed to a `dirty` state so we need to write some logic to know when that happens.

I'll try to implement this and see how far I get...

Reply:
Ok, it's sad... Anyway, they propose to use this method in the submit method:
```typescript
/**
   * Marks all controls in a form group as touched
   * @param formGroup - The form group to touch
   */
  private markFormGroupTouched(formGroup: FormGroup) {
    (<any>Object).values(formGroup.controls).forEach(control => {
      control.markAsTouched();

      if (control.controls) {
        this.markFormGroupTouched(control);
      }
    });
  }
```

You can see it here: https://github.com/angular/angular/issues/12281 

Because of this, I think that we should change this new line by 

```js
(this.pickerInputTouched || this.ngControl.dirty) &&
```

Like this, when the user clicks on "submit" button, the developer can simply mark the control as "dirty" and it will work correctly.
What do you think ? 😊

Reply:
Yes, indeed. In fact, the developer can mark the form control as `dirty` or `touched`, that is not the issue... the problem is to know when the control has been marked as `dirty` or `touched` so we can display the errors properly. Because knowing that the formFieldControl is in an "error state" is one thing and to display the errors is another thing.

I think I have an idea... I'll try to implement it

Reply:
If we check if it is "touched" (via `pickerInputTouched` variable) or check if the formControl is marked as touched (via `ngControl.touched`), the MatFormField will be in error state and the message will be displayed if it is defined in the mat-error if the the dev trigger the `updateValueAndValidity()` on the FormControl...

That means, we should adapt the method as this:
```typescript
/**
   * Marks all controls in a form group as touched
   * @param formGroup - The form group to touch
   */
private markFormGroupTouched(formGroup: FormGroup) {
  (<any>Object).values(formGroup.controls).forEach(control => {
    control.markAsTouched();
    control.updateValueAndValidity(); // <-- I added this line
    if (control.controls) {
      this.markFormGroupTouched(control);
    }
  });
}
```

and the line as this:
```js
(this.pickerInputTouched || !!this.ngControl.touched) &&
```

Otherwise, without the `ngControl.touched`, even if the developer set the formControl as invalid and touched, he won't be able to display the error message programatically since he cannot change the value of `pickerInputTouched`...

Reply:
I've found a simpler solution... which just requires the developer to mark the formControl as `dirty` or `touched`.

I'll update the PR 😉

Reply:
Ok, this is the solution I implemented:

```typescript
/**
 * Part of {@link MatFormFieldControl} API
 * @ignore
 * @internal
 */
public get errorState(): boolean {
	// the control can be in an error state as long as one of these conditions is met:
	// 1) the user has interacted with it
	// 2) the control is programmatically marked as 'touched' or 'dirty'
	const newErrorState =
		this.ngControl !== null &&
		this.ngControl.control !== null &&
		(this.pickerInputTouched || !!this.ngControl.touched || !!this.ngControl.dirty) &&
		(!!this.ngControl.errors || !!this.pickerInput.validate(this.ngControl.control));

	// IMPORTANT: emit an state change when the errorState changes
	// This is needed to force the MatFormFieldControl to refresh and render the MatError's
	if (this._errorState !== newErrorState) {
		this._errorState = newErrorState;
		this.stateChanges.next();
	}
		return this._errorState;
}

/**
 * The current error state
 * @ignore
 * @internal
 */
public _errorState = false;
```

So indeed, as you mentioned we need to check for `ngControl.touched` and `ngControl.dirty`... but the most important part to make the errors appear is to emit on the `stateChanges`.

Reply:
That seems to be a good solution :smiley: